modules/005-external-module-manager/hooks/migration_create_update_policies.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2023 Flant JSC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This migration hook creates module update policies for all but deckhouse module sources on first run.
+// After creation, d8-system namespace gets annotated and migration doesn't fire up anymore.
+package hooks
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/flant/addon-operator/pkg/module_manager/go_hook"
+	"github.com/flant/addon-operator/sdk"
+	"github.com/flant/shell-operator/pkg/kube/object_patch"
+	"github.com/iancoleman/strcase"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/deckhouse/deckhouse/deckhouse-controller/pkg/apis/deckhouse.io/v1alpha1"
+	"github.com/deckhouse/deckhouse/go_lib/dependency"
+)
+
+const (
+	defaultUpdateMode   = "Auto"
+	migrationAnnotation = "modules.deckhouse.io/ensured-update-policies"
+)
+
+var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	// ensure crds hook has order 5, for creating a module source we should use greater number
+	OnBeforeHelm: &go_hook.OrderedConfig{Order: 6},
+}, dependency.WithExternalDependencies(createModuleUpdatePolicies))
+
+var msResource = schema.GroupVersionResource{Group: "deckhouse.io", Version: "v1alpha1", Resource: "modulesources"}
+
+func createModuleUpdatePolicies(input *go_hook.HookInput, dc dependency.Container) error {
+	k8sCli, err := dc.GetK8sClient()
+	if err != nil {
+		return err
+	}
+
+	d8System, err := k8sCli.CoreV1().Namespaces().Get(context.TODO(), "d8-system", metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	// have already run the migration
+	if _, migrated := d8System.ObjectMeta.Annotations[migrationAnnotation]; migrated {
+		return nil
+	}
+
+	// get all modulesources
+	moduleSources, err := k8sCli.Dynamic().Resource(msResource).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, source := range moduleSources.Items {
+		// skip deckhouse source as its update policy is ensured by another hook
+		if source.GetName() == "deckhouse" {
+			continue
+		}
+		moduleSource := &v1alpha1.ModuleSource{}
+		err := sdk.FromUnstructured(&source, moduleSource)
+		if err != nil {
+			return err
+		}
+
+		// check if source releaseChannel can be camelCased correctly
+		rc, err := camelCaseReleaseChannel(moduleSource.Spec.ReleaseChannel)
+		if err != nil {
+			continue
```

Comment:
Lets add input.LogEntry.Warnf here, to notify that we skip a ModuleSource

Reply:
![Uploading image.png…]()